### PR TITLE
[SYCL] Return nullptr when allocation size is zero in usm allocator

### PR DESCRIPTION
**Pull request by uditagarwal97**
Currently, usm allocator throws when the allocation size is zero. However, this behavior is not aligned with that of std::allocator.
Refer https://github.com/KhronosGroup/SYCL-Docs/issues/355 for discussion regarding this. The spec says that the allocation functions must succeed when the size is zero.  The value returned in this case is unspecified (it can either be a NULL pointer or a non-NULL pointer)

This PR makes USM allocator return a null pointer when the allocation size is zero.

### CODE REVIEW

File: sycl/test-e2e/USM/usm_allocator_zero_allocation.cpp [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+using namespace sycl;
+#define ALLOC_SIZE 0
+
+int main() {
+  queue q;
+  auto dev = q.get_device();
+  auto ctxt = q.get_context();
+
+  if (dev.get_info<info::device::usm_host_allocations>()) {
```
Comment (aelovikov-intel) [inline sycl/test-e2e/USM/usm_allocator_zero_allocation.cpp:14]: Why do we only test "host" here?

File: sycl/test-e2e/USM/usm_allocator_zero_allocation.cpp [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    sycl::usm_allocator<int, sycl::usm::alloc::host> ua{ctxt, dev};
+    int *p = ua.allocate(ALLOC_SIZE);
+
+    assert(!p && "usm_allocator should return a null pointer when allocation "
+                 "size is zero.");
```
Comment (aelovikov-intel) [inline sycl/test-e2e/USM/usm_allocator_zero_allocation.cpp:19]: I'd reword as "our implementation is expected to ...".

File: sycl/test-e2e/USM/usm_allocator_zero_allocation.cpp (hunk 1 above)
Comment (uditagarwal97) [reply sycl/test-e2e/USM/usm_allocator_zero_allocation.cpp:23]: I'm not sure how adding more allocation types will increase the test coverage of this change. Nevertheless, I've expanded the test to also check for zero-sized shared allocations. usm::alloc::device is not supported by usm_allocator so, I've skipped that.

File: sycl/test-e2e/USM/usm_allocator_zero_allocation.cpp [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+using namespace sycl;
+#define ALLOC_SIZE 0
+
+int main() {
+  queue q;
+  auto dev = q.get_device();
+  auto ctxt = q.get_context();
+
+  if (dev.get_info<info::device::usm_host_allocations>()) {
+    sycl::usm_allocator<int, sycl::usm::alloc::host> ua{ctxt, dev};
```
[the remark this answers is not in the document]
Comment (uditagarwal97) [reply sycl/test-e2e/USM/usm_allocator_zero_allocation.cpp]: Fixed

File: sycl/test-e2e/USM/usm_allocator_zero_allocation.cpp [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+using namespace sycl;
+
+template <usm::alloc alloc_kind> void test(queue &q) {
+  sycl::usm_allocator<int, alloc_kind> ua(q);
+  int *p = ua.allocate(0);
+
+  assert(!p && "Our implementation of usm_allocator is expected to return a "
+               "null pointer when allocation "
+               "size is zero.");
```
Comment (gmlueck) [inline sycl/test-e2e/USM/usm_allocator_zero_allocation.cpp:14]: I think this is fine, but another option is to simply remove the assert.  From an API perspective, the important thing is that you can call allocate with a zero size and that you can then call deallocate on the returned pointer, all without any exception being raised.  It's not important whether or not the returned pointer is null.

**Review by gmlueck — APPROVED** [review]

File: sycl/test-e2e/USM/usm_allocator_zero_allocation.cpp (hunk 4 above)
Comment (aelovikov-intel) [inline sycl/test-e2e/USM/usm_allocator_zero_allocation.cpp:14]: Merge lines 13/14 into one.

File: sycl/test-e2e/USM/usm_allocator_zero_allocation.cpp [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+  ua.deallocate(p, 0);
+}
+
+int main() {
+  queue q;
+  auto dev = q.get_device();
+
+  if (dev.get_info<info::device::usm_host_allocations>()) {
```
Comment (aelovikov-intel) [inline sycl/test-e2e/USM/usm_allocator_zero_allocation.cpp:23]: I think it would be more idiomatic to `dev.has<aspect::usm_host_allocatinos>`.

File: sycl/test-e2e/USM/usm_allocator_zero_allocation.cpp [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    test<usm::alloc::host>(q);
+  }
+  if (dev.get_info<info::device::usm_shared_allocations>()) {
```
Comment (aelovikov-intel) [inline sycl/test-e2e/USM/usm_allocator_zero_allocation.cpp:26]: likewise - use aspect instead.

**Review by aelovikov-intel — APPROVED** [review]
